### PR TITLE
feat: adding tracing mechanism (MAPCO-3940)

### PR DESCRIPTION
**Pull request by ronenkapelian**
opentelemetry tracing mechanism + moving from zx to spawn

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

* first phase of adding minitoring features. 

* trace connected by link to parent trace (overseer)

* basic tracing.

* usage of the new semantic convention of raster and infra domain for trace attributes

* Convert usage of ZX library by implementing spawn to run cmd commands of mapproxy seed

* the next versions should add more metadata on each trace by future design

### CODE REVIEW

File: src/cacheSeed/cacheSeedManager.ts [hunk 1]
```diff
@@ -32,63 +36,95 @@ export class CacheSeedManager {
...
+  }
   public async delay(seconds: number): Promise<void> {
     this.logger.info(`waiting before executing by gracefulReloadRandomSeconds delay for -${seconds}- seconds `);
     await new Promise((resolve) => setTimeout(resolve, seconds * this.msToSeconds));
   }
 
+  // eslint-disable-next-line @typescript-eslint/member-ordering
+  @withSpanAsyncV4
   private async runTask(seedTasks: ISeed[], jobId: string, taskId: string): Promise<void> {
+    const spanActive = trace.getActiveSpan();
```
Comment (asafMasa) [inline src/cacheSeed/cacheSeedManager.ts:116]: can  you create a reusable method of get span options with parameters?

File: src/common/tracing.ts [hunk 2]
```diff
@@ -1,12 +1,35 @@
...
-]);
+const contextManager = new AsyncHooksContextManager();
+contextManager.enable();
+api.context.setGlobalContextManager(contextManager);
+
+export const tracing = new Tracing(
+  [new HttpInstrumentation({ requireParentforOutgoingSpans: true })],
+  {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    '@opentelemetry/instrumentation-express': { enabled: false },
```
Comment (asafMasa) [inline src/common/tracing.ts:17]: why?

File: tests/unit/testContainerConfig.ts [hunk 3]
```diff
@@ -8,8 +8,8 @@ import { InjectionObject } from '../../src/common/dependencyRegistration';
 import { IQueueConfig } from '../../src/common/interfaces';
 
 const queueConfig = configMock.get<IQueueConfig>('queue');
-tracing.start();
-const tracer = trace.getTracer('app');
+// tracing.start();
```
Comment (asafMasa) [inline tests/unit/testContainerConfig.ts:11]: why this is in comment?

**Review by asafMasa — CHANGES_REQUESTED** [review]
very good Job!

File: config/default.json [hunk 4]
```diff
@@ -3,6 +3,10 @@
     "logger": {
       "level": "info",
       "prettyPrint": false
+    },
+    "tracing": {
+      "enabled": true,
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply config/default.json:8]: I changed it, but according to Ofer it doesn't work on opentelemetry and only for visual, the actual activation provided by the environmental variable -TELEMETRY_TRACING_ENABLED=true npm run start

File: config/default.json [hunk 5]
```diff
@@ -42,6 +46,7 @@
     "geometryTxtFile": "/mapproxy/coverage.json",
     "seedProgressFileDir": "/mapproxy/mapproxy_seed_progress"
   },
+
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply config/default.json:49]: done

File: helm/values.yaml [hunk 6]
```diff
@@ -105,9 +109,6 @@ env:
   logPrettyPrintEnabled: false
   responseCompressionEnabled: true
   requestPayloadLimit: 1mb
-  tracing:
-    enabled: false
-    url: http://localhost:55681/v1/trace
   metrics:
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply helm/values.yaml:112]: added according to the custom variable file hierarchy 

image

File: src/cacheSeed/cacheSeedManager.ts [hunk 7]
```diff
@@ -32,63 +36,95 @@ export class CacheSeedManager {
...
+        }
 
+        // complete the current pool
+        return Boolean(tilesTask);
+      },
+      this.tracer,
+      'handleCacheSeedTask',
+      spanOptions
+    );
+  }
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:107]: done

File: src/cacheSeed/cacheSeedManager.ts (hunk 1 above)
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:116]: too complicated, with loadash manipulation and will hide some explicit content you want to provide the other developer

File: src/cacheSeed/cacheSeedManager.ts [hunk 8]
```diff
@@ -32,63 +36,95 @@ export class CacheSeedManager {
...
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'mapcolonies.raster.jobId': tilesTask.jobId,
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'mapcolonies.raster.taskId': tilesTask.id,
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'mapcolonies.raster.layerId': tilesTask.parameters.catalogId,
+        };
+      }
+    } catch (err) {
+      this.logger.warn({ msg: `No trace parent link data exists`, err: (err as Error).message });
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:53]: added

File: src/cacheSeed/cacheSeedManager.ts [hunk 9]
```diff
@@ -32,63 +36,95 @@ export class CacheSeedManager {
     if (!tilesTask) {
       return Boolean(tilesTask);
     }
-
-    const validCacheType = await this.isValidCacheType(tilesTask);
-    if (!validCacheType) {
-      return validCacheType;
+    let spanOptions = undefined;
+    try {
+      if (tilesTask.parameters.traceParentContext) {
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:41]: done.
have you checked empty traceParentContext object? --> checked

File: src/common/tracing.ts (hunk 2 above)
Comment (ronenkapelian) [reply src/common/tracing.ts:17]: the new configuration for optimal tracing according OFER, the other services use deprecated notations

File: tests/unit/testContainerConfig.ts (hunk 3 above)
Comment (ronenkapelian) [reply tests/unit/testContainerConfig.ts:11]: removed

File: src/mapproxyUtils/mapproxySeed.ts [hunk 10]
```diff
@@ -279,75 +378,15 @@ export class MapproxySeed {
         this.logger.info('requested to skip uncached tiles');
         flags.push('--skip-uncached');
       }
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      const cmdStr = `${this.mapproxyCmdCommand} ${flags.join(' ')}`;
+      this.logger.info({ msg: 'Execute cli command for seed', command: cmdStr });
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:382]: please add jobId, taskId

File: src/mapproxyUtils/mapproxySeed.ts [hunk 11]
```diff
@@ -106,24 +125,80 @@ export class MapproxySeed {
...
+  public seedProgressFunc(seedLogStr: string): void {
+    this.logger.debug(seedLogStr); // print all mapproxy-seed stdout
+    if (seedLogStr.match(/- ERROR -/g) && !isErroredCmd) {
+      // substr that detect seeding process error on mapproxy-seed util
+      // task will fail on case of seeding logic error (for example redis connection)
+      isErroredCmd = true;
+      const errMsg = seedLogStr.split('- ERROR -')[1];
+      this.logger.error(errMsg);
+      this.abortController.abort(errMsg);
+      console.log('0000');
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts]: console? - make sure no other console logs exists and change to "logger" when needed

File: src/common/cmd.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
+import { spawn } from 'node:child_process';
+
+let spawned = false;
+let stdErrStr = '';
+
+export async function runCommand(
```
Comment (asafMasa) [inline src/common/cmd.ts:6]: shouldn't it be in mc-utils for future reuse?

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Comment by asafMasa** [comment]
gracefulBumpFactor - where is it being used?

**Comment by asafMasa** [comment]
please fix conflicts

File: helm/values.yaml [hunk 13]
```diff
@@ -92,6 +92,10 @@ caSecretName: ''
 caPath: '/usr/local/share/ca-certificates'
 caKey: 'ca.crt'
 
+tracing:
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply helm/values.yaml:95]: moved

File: package.json [hunk 14]
```diff
@@ -46,21 +46,21 @@
...
     "express": "^4.18.2",
-    "js-yaml": "^4.1.0",
-    "moment": "^2.29.4",
     "http-status-codes": "^2.2.0",
+    "js-yaml": "^4.1.0",
     "lodash": "^4.17.21",
+    "moment": "^2.29.4",
     "reflect-metadata": "^0.1.13",
     "tsyringe": "^4.8.0",
+    "zod": "^3.22.4",
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply package.json:63]: removed

File: src/cacheSeed/cacheSeedManager.ts [hunk 15]
```diff
@@ -122,4 +103,77 @@ export class CacheSeedManager {
...
     }
   }
+
+  /**
+   * This function parse seedTask and generate SpanOption object to be passed, attach Link object to Span parent if exists, and metadata attributes
+   * @param tilesTask seed task to be executed
+   * @returns SpanOption object with attributes and optional links array
+   */
+  private getInitialSpanOption(tilesTask: ITaskResponse<ITaskParams>): SpanOptions {
+    let spanLinks;
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:113]: you right - i changed to const

File: src/cacheSeed/cacheSeedManager.ts [hunk 16]
```diff
@@ -122,4 +103,77 @@ export class CacheSeedManager {
...
+    };
+    try {
+      if (tilesTask.parameters.traceParentContext) {
+        spanLinks = getSpanLinkOption(tilesTask.parameters.traceParentContext); // add link to trigging parent trace (overseer)
+        spanOptions.links = spanLinks;
+      }
+    } catch (err) {
+      const logWarnMsg = `No trace parent link data exists`;
+      const logObj = { jobId: tilesTask.jobId, taskId: tilesTask.id, layerId: tilesTask.parameters.catalogId, err: (err as Error).message };
+      this.logger.warn(logObj, logWarnMsg);
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:130]: changed as we discussed

File: src/cacheSeed/cacheSeedManager.ts [hunk 17]
```diff
@@ -122,4 +103,77 @@ export class CacheSeedManager {
...
+    try {
+      await this.delay(this.gracefulReloadMaxSeconds);
+      await this.runTask(seeds, jobId, taskId);
+      await this.queueClient.queueHandlerForTileSeedingTasks.ack(jobId, taskId);
+      await this.queueClient.queueHandlerForTileSeedingTasks.jobManagerClient.updateJob(jobId, {
+        status: OperationStatus.COMPLETED,
+        percentage: 100,
+      });
+    } catch (error) {
+      await this.queueClient.queueHandlerForTileSeedingTasks.reject(jobId, taskId, true, (error as Error).message);
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:171]: added

File: src/clients/mapproxyConfig.ts [hunk 18]
```diff
@@ -20,13 +27,23 @@ export class MapproxyConfigClient extends HttpClient {
     );
   }
 
+  @withSpanAsyncV4
   public async getConfig(format: SchemaType = SchemaType.YAML): Promise<IMapProxyConfig | Promise<string>> {
+    const spanActive = trace.getActiveSpan();
+    spanActive?.setAttributes({
+      /* eslint-disable @typescript-eslint/naming-convention */
+      'mapcolonies.raster.mapproxyApiUrl': this.getConfigUrl,
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:35]: REMOVED

File: src/clients/mapproxyConfig.ts [hunk 19]
```diff
@@ -20,13 +27,23 @@ export class MapproxyConfigClient extends HttpClient {
...
+      /* eslint-enable @typescript-eslint/naming-convention */
+    });
     // eslint-disable-next-line @typescript-eslint/naming-convention
     const headers = { Accept: format };
     try {
-      this.logger.info({ msg: `Getting mapproxy.yaml configuration from provider request` });
-      const mapproxyConfig: IMapProxyConfig | string = await this.get(this.getConfigUr, undefined, undefined, undefined, headers);
+      const logInfoMsg = `Getting mapproxy.yaml configuration from provider request`;
+      const logInfoObj = { mapproxyApiUrl: join([this.config.get<string>('mapproxy.mapproxyApiUrl'), this.getConfigUrl], '') };
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:42]: done

File: src/clients/mapproxyConfig.ts [hunk 20]
```diff
@@ -20,13 +27,23 @@ export class MapproxyConfigClient extends HttpClient {
...
+      this.logger.info(logInfoObj, logInfoMsg);
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:43]: no need. it autonomic request for the current mapproxy config. I don't see a reason break the function definition for it

File: src/clients/mapproxyConfig.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:42]: but not in ctor, this needed only for log info , the actual construct happen with HTTP client

File: src/common/cmd.ts (hunk 12 above)
Comment (ronenkapelian) [reply src/common/cmd.ts:6]: yes, but I prefer as single task on future after some time of using and see evaluation

File: src/mapproxyUtils/mapproxySeed.ts (hunk 10 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:382]: done

File: src/cacheSeed/cacheSeedManager.ts [hunk 21]
```diff
@@ -168,6 +167,8 @@ export class CacheSeedManager {
         percentage: 100,
       });
     } catch (error) {
+      const errorObj = { jobId, taskId, msg: 'Reject task and increase attempts', error };
```
Comment (asafMasa) [inline src/cacheSeed/cacheSeedManager.ts:170]: i think all errors should be "error" or "err" can't be both.
in error log you can pass the error object separatably as is done in all services

File: src/clients/mapproxyConfig.ts [hunk 22]
```diff
@@ -30,17 +30,13 @@ export class MapproxyConfigClient extends HttpClient {
...
     const spanActive = trace.getActiveSpan();
-    spanActive?.setAttributes({
-      /* eslint-disable @typescript-eslint/naming-convention */
-      'mapcolonies.raster.mapproxyApiUrl': this.getConfigUrl,
-      /* eslint-enable @typescript-eslint/naming-convention */
-    });
     // eslint-disable-next-line @typescript-eslint/naming-convention
     const headers = { Accept: format };
     try {
+      const mapproxyApiUrl = join([this.config.get<string>('mapproxy.mapproxyApiUrl'), this.getConfigUrl], '');
```
Comment (asafMasa) [inline src/clients/mapproxyConfig.ts:36]: why do you need the join method to be activated every time this method is called and not just once in constructor?

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/clients/mapproxyConfig.ts (hunk 22 above)
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:36]: it's for logging, the actual concat occurs inside the HTTP client class (hidden from the developer)

File: src/cacheSeed/cacheSeedManager.ts (hunk 21 above)
Comment (ronenkapelian) [reply src/cacheSeed/cacheSeedManager.ts:170]: done

File: src/clients/mapproxyConfig.ts (hunk 22 above)
Comment (asafMasa) [reply src/clients/mapproxyConfig.ts:36]: as discussed

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: helm/templates/configmap.yaml [hunk 23]
```diff
@@ -14,13 +14,13 @@ data:
...
-  {{ if .Values.env.tracing.enabled }}
-  TELEMETRY_TRACING_ENABLED: 'true'
-  TELEMETRY_TRACING_URL: {{ $tracingUrl }}
+  TELEMETRY_TRACING_ENABLED: {{ $tracing.enabled | quote }}
+  {{ if $tracing.enabled }}
+  TELEMETRY_TRACING_URL: {{ $tracing.url }}
   {{ end }}
-  {{ if .Values.env.metrics.enabled }}
+  {{ if $metrics.enabled }}
   TELEMETRY_METRICS_ENABLED: 'true'
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply helm/templates/configmap.yaml:22]: done - you right

File: helm/values.yaml [hunk 24]
```diff
@@ -93,6 +93,14 @@ caSecretName: ''
...
+  enabled: false
+  url: ''
+
+metrics:
+    enabled: false
+    url: http://localhost:55681/v1/metrics
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply helm/values.yaml:102]: done + fix indentation

File: src/clients/mapproxyConfig.ts (hunk 22 above)
Comment (ronenkapelian) [reply src/clients/mapproxyConfig.ts:36]: done

**Review by asafMasa — APPROVED** [review]

**Pull request merged**